packages/compiler/core/program.ts

```diff
@@ -418,18 +420,33 @@ export async function createProgram(
     if (program.sourceFiles.has(cadlScript.path)) {
       throw new RangeError("Duplicate script path: " + cadlScript);
     }
-    const sourceFile = parse(cadlScript);
+    const sourceFile = parseOrReuse(cadlScript);
     program.reportDiagnostics(sourceFile.parseDiagnostics);
     program.sourceFiles.set(cadlScript.path, sourceFile);
     for (const stmt of sourceFile.statements) {
       if (stmt.kind !== SyntaxKind.ImportStatement) break;
       program.reportDiagnostic(createDiagnostic({ code: "dynamic-import", target: stmt }));
     }
     binder.bindSourceFile(sourceFile);
-
     return sourceFile;
   }
 
+  function parseOrReuse(file: SourceFile): CadlScriptNode {
+    if (!host.canReuseFiles) {
+      return parse(file);
+    }
+
+    parseAndBindCache ??= new Map();
+    const cached = parseAndBindCache.get(file);
```

Comment:
how does this get invalidated, I don't really see how it gets connected in the serverlib

Reply:
It's a weakmap, so if the SourceFile gets collected, it goes away.

Reply:
The serverlib will return existing SourceFile instances and then this will find them in the weak map. I am not sure this is the best approach, though.

Reply:
Lol, new Map(). It was meant to be a WeakMap. Sorry.

Reply:
And.... Map is assignable to WeakMap. Yay, structural typing.

Reply:
oh I see, SourceFile being a new instance everytime there is a change. make sense